python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py

```diff
@@ -362,79 +390,128 @@ def _prepare_settings(
             )
             chat_history.add_message(message=frc.to_chat_message_content())
             return
-        logger.info(f"Calling {func.name} function with args: {func.arguments}")
+
+        logger.info(f"Calling {function_call.name} function with args: {function_call.arguments}")
         try:
-            func_result = await kernel.invoke(**func.split_name_dict(), arguments=args_cloned)
+            if function_call.name is None:
+                raise ValueError("The function name is required.")
+            if isinstance(function_call_behavior, RequiredFunction):
+                if function_call.name != function_call_behavior.function_fully_qualified_name:
+                    raise ValueError(
+                        f"Only function: {function_call_behavior.function_fully_qualified_name} "
+                        f"is allowed, {function_call.name} is not allowed."
+                    )
+            if isinstance(function_call_behavior, EnabledFunctions):
+                enabled_functions = [
+                    func.fully_qualified_name
+                    for func in kernel.get_list_of_function_metadata(function_call_behavior.filters)
+                ]
+                if function_call.name not in enabled_functions:
+                    raise ValueError(
+                        f"Only functions: {enabled_functions} are allowed, {function_call.name} is not allowed."
+                    )
+            function_to_call = kernel.get_function(function_call.plugin_name, function_call.function_name)
         except Exception as exc:
-            logger.exception(f"Exception occurred while invoking function {func.name}, exception: {exc}")
+            logger.exception(f"Could not find function {function_call.name}: {exc}.")
             frc = FunctionResultContent.from_function_call_content_and_result(
-                function_call_content=result,
-                result=f"Exception occurred while invoking function {func.name}, exception: {exc}",
+                function_call_content=function_call,
+                result="The tool call could not be found, please try again and make sure to validate the name.",
             )
             chat_history.add_message(message=frc.to_chat_message_content())
             return
+
+        _rebuild_auto_function_invocation_context()
+        invocation_context = AutoFunctionInvocationContext(
+            function=function_to_call,
+            kernel=kernel,
+            arguments=args_cloned,
+            chat_history=chat_history,
+            function_result=FunctionResult(function=function_to_call.metadata, value=None),
+            function_count=function_call_count,
+            request_sequence_index=request_index,
+        )
+        if function_call.index is not None:
+            invocation_context.function_sequence_index = function_call.index
+
+        stack = kernel.construct_call_stack(
+            filter_type="auto_function_invocation",
```

Comment:
Would it be better to have this as a Pydantic literal or an enum somewhere to avoid typing mistakes?